File: pennywise-ui/src/types/transaction.ts
export enum TransactionType {
  INCOME = "INCOME",
  EXPENSE = "EXPENSE"
}

export interface Transaction {
  id: number;
  group_id: number;
  user_id: number;
  amount: number;
  type: TransactionType;
  note?: string;
  category?: string;
  payment_mode?: string;
  date: string;
  paid_by?: number;
  // User information
  user_full_name?: string;
  user_email?: string;
  user_username?: string;
  // Paid by user information
  paid_by_full_name?: string;
  paid_by_email?: string;
  paid_by_username?: string;

  group_name?: string;
}

export interface TransactionCreate {
  group_id: number;
  user_id: number;
  amount: number;
  type: TransactionType;
  note?: string;
  category?: string;
  payment_mode?: string;
  date?: string;
  paid_by?: number;
}

export interface PaginatedTransactionResponse {
  transactions: Transaction[];
  total: number;
  skip: number;
  limit: number;
  has_more: boolean;
}

File: pennywise-backend/app/schemas/transaction.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from app.constants.transactions import TransactionType

class TransactionBase(BaseModel):
    group_id: int  # Refers to Group
    user_id: int   # Refers to User
    amount: float
    type: TransactionType
    note: Optional[str] = None
    category: Optional[str] = None
    payment_mode: Optional[str] = None
    date: Optional[datetime] = None
    paid_by: Optional[int] = None  # User ID who paid

class TransactionCreate(TransactionBase):
    pass

class TransactionUpdate(BaseModel):
    """Complete transaction object for updates - includes all fields"""
    id: int
    group_id: int
    user_id: int
    amount: float
    type: TransactionType
    note: Optional[str] = None
    category: Optional[str] = None
    payment_mode: Optional[str] = None
    date: Optional[datetime] = None
    paid_by: Optional[int] = None
    # User information (optional for updates)
    user_full_name: Optional[str] = None
    user_email: Optional[str] = None
    user_username: Optional[str] = None
    # Paid by user information (optional for updates)
    paid_by_full_name: Optional[str] = None
    paid_by_email: Optional[str] = None
    paid_by_username: Optional[str] = None

class BulkTransactionCreate(BaseModel):
    transactions: List[TransactionCreate]

class TransactionResponse(TransactionBase):
    id: int
    date: Optional[datetime] = None
    # User information
    user_full_name: Optional[str] = None
    user_email: Optional[str] = None
    user_username: Optional[str] = None
    # Paid by user information
    paid_by_full_name: Optional[str] = None
    paid_by_email: Optional[str] = None
    paid_by_username: Optional[str] = None

    class Config:
        from_attributes = True

class PaginatedTransactionResponse(BaseModel):
    transactions: List[TransactionResponse]
    total: int
    skip: int
    limit: int
    has_more: bool


# AI Transaction Extraction Schemas
class TransactionExtractRequest(BaseModel):
    """Request payload for transaction extraction."""
    text: str = Field(..., description="Free-form text to extract transactions from.")


class TransactionExtractResponse(BaseModel):
    """Response containing extracted transactions."""
    transactions: List[dict] = Field(..., description="List of extracted transactions.")
    total_count: int = Field(..., description="Total number of transactions extracted.") 